Additional EWR features

emrahkocaman opened the pull request:
oss part of https://github.com/hazelcast/hazelcast-enterprise/issues/588

emrahkocaman commented:
run-lab-run

enesakar commented:
@pveentjer can you review this PR asap? we need this for EA3 release

bilalyasar commented on hazelcast/src/main/java/com/hazelcast/config/XmlConfigBuilder.java:
```diff
@@ -500,6 +480,46 @@ private void handleWanReplication(final Node node) throws Exception {
+            long batchMaxDelayMillis = getLongValue("batch-max-delay-millis", getTextContent(targetChild));
+            wanTarget.setBatchMaxDelayMillis(batchMaxDelayMillis);
+        } else if ("queue-capacity".equals(targetChildName)) {
+            int queueCapacity = getIntegerValue("queueCapacity", getTextContent(targetChild));
```
`getIntegerValue("queueCapacity"...` or getIntegerValue("queue-capacity"... ?

emrahkocaman replied:
Should be `queue-capacity`

bilalyasar commented on hazelcast/src/main/java/com/hazelcast/config/XmlConfigBuilder.java:
```diff
@@ -500,6 +480,46 @@ private void handleWanReplication(final Node node) throws Exception {
+    private void handleWanTargetConfig(WanReplicationConfig wanReplicationConfig,
+                                       WanTargetClusterConfig wanTarget, Node targetChild) {
+        final String targetChildName = cleanNodeName(targetChild);
+        if ("replication-impl".equals(targetChildName)) {
```
what if there is no class, fail-fast? what do you think @mesutcelik ?

mesutcelik replied:
we can set a default version....

emrahkocaman replied:
Node start will fail with `ClassNotFoundException` in that case. But I think this part is out of scope of this since there is nothing new here. I just extracted them to a method logic is same as before.

mesutcelik replied:
I think that is generic problem with all configs that accept fully qualified class names... @bilalyasar , can you please create an issue for that?

mesutcelik commented on hazelcast/src/main/java/com/hazelcast/config/WanQueueFullBehavior.java:
```diff
@@ -0,0 +1,65 @@
+     *
+     * @return the {@link WanQueueFullBehavior} found or null if not found
+     */
+    public static WanQueueFullBehavior getById(final int id) {
```
getById seems not used...

emrahkocaman replied:
In fact, ID is totally unnecessary since this config is not serialized. I'll remove.

mesutcelik commented on hazelcast/src/main/java/com/hazelcast/config/WanQueueFullBehavior.java:
```diff
@@ -0,0 +1,65 @@
+    /**
+     * Instruct WAN repl. impl to drop new events when WAN event queues are full
+     */
+    DISCARD_AFTER_MUTATION(0),
```
what does `DISCARD_AFTER_MUTATION` mean?

emrahkocaman replied:
Actually this is the existing behavior of WAN replication and will be default behavior of the new version also.
When `DISCARD_AFTER_MUTATION` is chosen, mutating operation (put/remove/etc...) will be applied but WAN event will be discarded if WAN replication queue is full.

When `THROW_EXCEPTION` is chosen, a queue size check will be performed in `beforeRun` of mutating operations and `WANReplicationQueueFullException` will be throwed if WAN replication queue is full, therefore mutating operation (put/remove/etc...) won't run.

mesutcelik replied:
ok

mesutcelik commented on hazelcast/src/main/java/com/hazelcast/config/WanTargetClusterConfig.java:
```diff
@@ -24,6 +24,16 @@
+    private static final int DEFAULT_BATCH_SIZE = 500;
+    private static final long DEFAULT_BATCH_MAX_DELAY_MILLIS = 1000;
+    private static final int DEFAULT_QUEUE_CAPACITY = 10000;
+    private static final long DEFAULT_RESPONSE_TIMEOUT_MILLIS = 60000;
```
I see those final statics not used anywhere else so what is the point to declare them here and then assing to class fields?

emrahkocaman replied:
Checkstyle doesn't allow magic numbers.

mesutcelik replied:
ok. I would prefer to create a constructor which is accepting all class fields and remove all setters. Default constructor would use those static fields. However, that is my preference...

mesutcelik commented on hazelcast/src/main/java/com/hazelcast/map/impl/operation/PutAllOperation.java:
```diff
@@ -68,6 +68,14 @@ public PutAllOperation(String name, MapEntries mapEntries, boolean initialLoad)
     @Override
+    public void beforeRun() throws Exception {
+        super.beforeRun();
+        if (shouldWanReplicate()) {
```
What is the diff between `mapContainer.isWanReplicationEnabled()` and `shouldWanReplicate`?

emrahkocaman replied:
Seems like they are same. It's possibly a leftover from times before `isWanReplicationEnabled` method is introduced.

emrahkocaman replied:
I'll remove `shouldWanReplicate` method.

serkan-ozal commented on hazelcast/src/main/java/com/hazelcast/cache/impl/operation/CacheGetAndRemoveOperation.java:
```diff
@@ -37,6 +37,13 @@ public CacheGetAndRemoveOperation(String name, Data key, int completionId) {
 
     @Override
+    protected void beforeRunInternal() {
+        if (cache.isWanReplicationEnabled()) {
```
not important but `super.beforeRunInternal();` is needed here also?

emrahkocaman replied:
:+1: I think so.

ahmetmircik commented on hazelcast/src/main/java/com/hazelcast/map/impl/operation/AbstractMultipleEntryOperation.java:
```diff
@@ -67,6 +67,15 @@ protected AbstractMultipleEntryOperation(String name, EntryBackupProcessor backu
+        super.beforeRun();
+        if (mapContainer.isWanReplicationEnabled()) {
+            mapContainer.checkWanReplicationQueues();
+        }
```
We have a [MapOperation] (https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapOperation.java#L52) to centralize this kind of checks.

emrahkocaman replied:
But we need to make this check only in mutating operations. Otherwise, for example a get operation  will also throw exception when WAN queue is full.

emrahkocaman commented:
I've updated the PR.

For map, @ahmetmircik, I've moved WAN checks to `WANAwareOperationProvider` and cleaned `beforeRun` methods.

For cache, @serkan-ozal , Also cleaned up WAN related checks from cache operations. They'll be available in ee.

Also I've added tests for configuration part.

Guys, can you please check again?

ahmetmircik commented on hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapOperationProviders.java:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    /**
+     * Creates {@link MapOperationProvider} instance and wraps it into a {@link WANAwareOperationProvider} if
```
`if` seems leftover in the end.

emrahkocaman replied:
:+1:

ahmetmircik commented on hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapOperationProviderDelegator.java:
```diff
@@ -0,0 +1,236 @@
+ */
+public abstract class MapOperationProviderDelegator implements MapOperationProvider {
+
+    abstract MapOperationProvider getOperationProviderDelegate();
```
Maybe renamed as `getDelegate()`, we know this is map-operation-provider from context.

emrahkocaman replied:
I'll update.

ahmetmircik commented:
:+1:  with minor comments.

emrahkocaman commented:
run-lab-run

mesutcelik commented:
:+1:
